design/client-side-encryption_design.md

```diff
@@ -0,0 +1,554 @@
+                backupErr = err
+            }
+        }()
+    } else {
+        backupContent = backupFile
+    }
+
+    gzippedData := gzip.NewWriter(backupContent)
+    ...
+}
+```
+
+### Decryption on restore
+
+The `kubernetesRestorer` gets extended by a namespace:
+```go
+type kubernetesRestorer struct {
+    ...
+    veleroNamespace string
+}
+
+func NewKubernetesRestorer(
+    ...
+    veleroNamespace string,
+) (Restorer, error) {
+    ...
+}
+```
+
+The namespace is the namespace velero runs in.
+It is necessary to find the encryption secret.
+
+The namespace also needs to be passed to `restoreContext`:
+```go
+type restoreContext struct {
+    ...
+    veleroNamespace string
+}
+```
+
+Decryption then happens in the `execute` method of the `restoreContext`:
+```go
+func (ctx *restoreContext) execute() (results.Result, results.Result) {
+    warnings, errs := results.Result{}, results.Result{}
+
+    ctx.log.Infof("Starting restore of backup %s", kube.NamespaceAndName(ctx.backup))
+
+    var backupContent io.Reader
+    var err error
+    if ctx.backup.Status.Encryption.IsEncrypted {
+    encryptionKeyRetriever, err := encryption.KeyRetrieverFor(
+        ctx.backup.Status.Encryption.KeyRetrieverType,
+        ctx.backup.Status.Encryption.KeyRetrieverConfig,
+        ctx.kbClient,
+    )
+    if err != nil {
+        ctx.log.Errorf("error creating encryption key retriever: %s", err.Error())
+        errs.AddVeleroError(err)
+        return warnings, errs
+    }
+
+    encryptionKey, err := encryptionKeyRetriever.GetKey()
+    if err != nil {
+        ctx.log.Errorf("error getting encryption key from secret: %s", err.Error())
+        errs.AddVeleroError(err)
+        return warnings, errs
+    }
+
+    backupContent, err = encryption.NewDecryptionReader(ctx.backupReader, encryptionKey)
+    if err != nil {
+        ctx.log.Errorf("error decrypting backup: %s", err.Error())
+        errs.AddVeleroError(err)
+        return warnings, errs
+    }
+
+    } else {
+        backupContent = ctx.backupReader
+    }
+
+    dir, err := archive.NewExtractor(ctx.log, ctx.fileSystem).UnzipAndExtractBackup(backupContent)
+    ...
+}
+```
+
+## Alternatives Considered
+
+There are other ways this could be implemented.
+One valid way would be to do it on a per-backup basis instead of globally.
+This could be implemented by adding the fields `isEncrypted` and `encryptionSecret` to the backup spec instead of its status.
+An advantage of this would be that there is no need for a feature flag  or the `encryptionSecret` flag anymore.
+On the other hand, it would be fatal if a user forgets to encrypt the backup.
+But as most users will likely automate their backups anyway, this shouldn't pose that much of a problem.
+
+This could also be implemented in object store plugins.
+Then however, it has to be implemented in every single object store plugin.
+This is not viable.
+
+We could just make use of server-side encryption of an object store.
```

Comment:
most cloud based Object stores  do support encryption at rest using user managed keys.

Can you give production examples where users are using object stores which don't support such functionality.

Reply:
Data should also be encrypted in transit using TLS, though having client side encryption is useful as well.

Reply:
A case where this may help is if bsl is set to use insecure for whatever reason.. ie. User decides to use aws plugin and override s3url to use a minio server which is uses self signed cert and have not bothered to setup cacert properly.